File: src/Test.CSharp/Test7/Test7.cs
using System;
using System.Threading;
using SecretNest.RemoteAgency;
using SecretNest.RemoteAgency.Attributes;

namespace Test.CSharp.Test7
{
    public interface ITest7
    {
        [AssetIgnored]
        DateTime Ignored { get; set; }

        [AssetOneWayOperating]
        DateTime OneWaySet { get; set; }

        [PropertyGetOneWayOperating]
        DateTime OneWayGet { get; }

        [OperatingTimeoutTime(1000, 2000)]
        DateTime TimeoutExceptionTest { get; set; }

        [LocalExceptionHandling]
        EntityInTest7 WithException { get; set; }
    }

    public class EntityInTest7
    {
        public string FromClientToServerProperty { get; set; }

        [ParameterReturnRequiredProperty("EntityTwoWayProperty", isIncludedInReturning: true)]
        public string TwoWayProperty { get; set; }
    }

    public class Server7 : ITest7
    {
        public DateTime Ignored {
            get => throw new Exception("You should never see this due to ignored.");
            set => throw new Exception("You should never see this due to ignored.");
        }

        public DateTime OneWaySet
        {
            get => DateTime.Now;
            set
            {
                Console.WriteLine("Server side: Operation received. Value={0}", value.ToLongDateString());
                throw new Exception("You should never receive this from client coz this is a one way (set) property.");
            }
        }

        public DateTime OneWayGet
        {
            get
            {
                Console.WriteLine("Server side: Operation received.");
                throw new Exception("You should never receive this from client coz this is a one way (set) property.");
            }
        }

        public DateTime TimeoutExceptionTest
        {
            get
            {
                Thread.Sleep(2000);
                return DateTime.Now;
            }
            set
            {
                Thread.Sleep(3000);
                _ = value;
            }
        }

        public EntityInTest7 WithException
        {
            get => null;
            set
            {
                value.FromClientToServerProperty = "ChangedByServer"; //should never returned
                value.TwoWayProperty = "SetBeforeException";
                throw new Exception("oops.");
            }
        }
    }

    public static class TestCode
    {
        public static void MyTest()
        {
            //test router
            var router = new RemoteAgencyRouter<byte[], object>();
            
            //Server
            var originalService = new Server7();
            using var serverRemoteAgencyInstance = RemoteAgencyBase.CreateWithBinarySerializer(true);
            router.AddRemoteAgencyInstance(serverRemoteAgencyInstance);
            var serverSiteId = serverRemoteAgencyInstance.SiteId;
            var serviceWrapperInstanceId = serverRemoteAgencyInstance.CreateServiceWrapper(originalService);

            //Client
            using var clientRemoteAgencyInstance = RemoteAgencyBase.CreateWithBinarySerializer(true);
            router.AddRemoteAgencyInstance(clientRemoteAgencyInstance);
            var clientProxy = clientRemoteAgencyInstance.CreateProxy<ITest7>(serverSiteId, serviceWrapperInstanceId).ProxyGeneric;
            clientRemoteAgencyInstance.ExceptionRedirected += ClientRemoteAgencyInstance_ExceptionRedirected;

            //Run test
            Console.WriteLine("Ignored(Exception):");
            try
            {
                _ = clientProxy.Ignored;
            }
#pragma warning disable CA1031 // Do not catch general exception types
            catch (Exception e)
            {
                Console.WriteLine("Predicted Exception: " + e);
            }
#pragma warning restore CA1031 // Do not catch general exception types

            Console.WriteLine("OneWaySet(NoClientException):");
            clientProxy.OneWaySet = DateTime.Now;

            Console.WriteLine("OneWayGet(default):");
            Console.WriteLine(clientProxy.OneWayGet);

            Console.WriteLine("TimeoutExceptionTest:");
            try
            {
                Console.WriteLine(clientProxy.TimeoutExceptionTest);
            }
#pragma warning disable CA1031 // Do not catch general exception types
            catch (Exception e)
            {
                Console.WriteLine("Predicted Exception: " + e);
            }
#pragma warning restore CA1031 // Do not catch general exception types

            var entity = new EntityInTest7
            {
                FromClientToServerProperty = "SetFromClient",
                TwoWayProperty = "SetFromClient"
            };

            Console.WriteLine("WithException(Exception):");
            try
            {
                clientProxy.WithException = entity;
            }
#pragma warning disable CA1031 // Do not catch general exception types
            catch (Exception e)
            {
                Console.WriteLine("Predicted Exception: " + e.Message);
            }
#pragma warning restore CA1031 // Do not catch general exception types
            Console.WriteLine($"Client side: entity.FromClientToServerProperty (should be SetFromClient): {entity.FromClientToServerProperty}");
            Console.WriteLine($"Client side: entity.TwoWayProperty (should be SetBeforeException): {entity.TwoWayProperty}");

            Console.Write("Press any key to continue...");
            Console.ReadKey(true);
            Console.WriteLine();
        }

        private static void ClientRemoteAgencyInstance_ExceptionRedirected(object sender, ExceptionRedirectedEventArgs e)
        {   
            Console.WriteLine($"Client side received exception: \n  Interface:{e.ServiceContractInterface.FullName}\n  InstanceId: {e.InstanceId}\n  AssetName: {e.AssetName}\n  ExceptionType: {e.RedirectedException.GetType().FullName}\n  ExceptionMessage: {e.RedirectedException.Message}");
        }
    }
}

File: src/Test.VB/Test7/Test7.vb
Imports System.Threading
Imports SecretNest.RemoteAgency
Imports SecretNest.RemoteAgency.Attributes

Namespace Test7
    Public Interface ITest7
        <AssetIgnored>
        Property Ignored As DateTime

        <AssetOneWayOperating>
        Property OneWaySet As DateTime

        <PropertyGetOneWayOperating>
        ReadOnly Property OneWayGet As DateTime

        <OperatingTimeoutTime(1000, 2000)>
        Property TimeoutExceptionTest As DateTime

        <LocalExceptionHandling>
        Property WithException As EntityInTest7
    End Interface

    Public Class EntityInTest7
        Public Property FromClientToServerProperty As String

        <ParameterReturnRequiredProperty("EntityTwoWayProperty", Nothing, True)>
        Public Property TwoWayProperty As String
    End Class

    Public Class Server7
        Implements ITest7

        Public Property Ignored As Date Implements ITest7.Ignored
            Get
                Throw New Exception("You should never see this due to ignored.")
            End Get
            Set(value As Date)
                Throw New Exception("You should never see this due to ignored.")
            End Set
        End Property

        Public Property OneWaySet As Date Implements ITest7.OneWaySet
            Get
                Return DateTime.Now
            End Get
            Set(value As Date)
                Console.WriteLine("Server side: Operation received. Value={0}", value.ToLongDateString())
                Throw New Exception("You should never receive this from client coz this is a one way (set) property.")
            End Set
        End Property

        Public ReadOnly Property OneWayGet As Date Implements ITest7.OneWayGet
            Get
                Console.WriteLine("Server side: Operation received.")
                Throw New Exception("You should never receive this from client coz this is a one way (get) property.")
            End Get
        End Property

        Public Property TimeoutExceptionTest As Date Implements ITest7.TimeoutExceptionTest
            Get
                Thread.Sleep(2000)
                Return DateTime.Now
            End Get
            Set(value As Date)
                Thread.Sleep(3000)
            End Set
        End Property

        Public Property WithException As EntityInTest7 Implements ITest7.WithException
            Get
                Return Nothing
            End Get
            Set(value As EntityInTest7)
                value.FromClientToServerProperty = "ChangedByServer" 'should never returned
                value.TwoWayProperty = "SetBeforeException"
                Throw New Exception("oops.")
            End Set
        End Property
    End Class

    Public NotInheritable Class TestCode
        Public Shared Sub MyTest()
            'test router
            Dim router = New RemoteAgencyRouter(Of Byte(), Object)

            'Server
            Dim originalService As New Server7()
            Dim serverRemoteAgencyInstance = RemoteAgencyBase.CreateWithBinarySerializer(True)
            router.AddRemoteAgencyInstance(serverRemoteAgencyInstance)
            AddHandler serverRemoteAgencyInstance.ExceptionRedirected, AddressOf ServerRemoteAgencyInstance_ExceptionRedirected
            Dim serverSiteId = serverRemoteAgencyInstance.SiteId
            Dim serviceWrapperInstanceId = serverRemoteAgencyInstance.CreateServiceWrapper(originalService)

            'Client
            Dim clientRemoteAgencyInstance = RemoteAgencyBase.CreateWithBinarySerializer(True)
            router.AddRemoteAgencyInstance(clientRemoteAgencyInstance)
            Dim clientProxy = clientRemoteAgencyInstance.CreateProxy(Of ITest7)(serverSiteId, serviceWrapperInstanceId).ProxyGeneric

            'Run test
            Console.WriteLine("Ignored(Exception):")
            Try
                ' ReSharper disable once UnusedVariable
                Dim useless = clientProxy.Ignored
#Disable Warning CA1031 ' Do not catch general exception types
            Catch ex As Exception
                Console.WriteLine("Predicted Exception: " + ex.ToString())
            End Try
#Enable Warning CA1031 ' Do not catch general exception types

            Console.WriteLine("OneWaySet(NoClientException):")
            clientProxy.OneWaySet = DateTime.Now

            Console.WriteLine("OneWayGet(default):")
            Console.WriteLine(clientProxy.OneWayGet)

            Console.WriteLine("TimeoutExceptionTest:")
            Try
                Console.WriteLine(clientProxy.TimeoutExceptionTest)
#Disable Warning CA1031 ' Do not catch general exception types
            Catch ex As Exception
                Console.WriteLine("Predicted Exception: " + ex.ToString())
            End Try
#Enable Warning CA1031 ' Do not catch general exception types

            Dim entity As New EntityInTest7()
            With entity
                .FromClientToServerProperty = "SetFromClient"
                .TwoWayProperty = "SetFromClient"
            End With

            Console.WriteLine("WithException(Exception):")
            Try
                clientProxy.WithException = entity
#Disable Warning CA1031 ' Do not catch general exception types
            Catch ex As Exception
                Console.WriteLine("Predicted Exception: " + ex.Message)
            End Try
#Enable Warning CA1031 ' Do not catch general exception types
            Console.WriteLine($"Client side: entity.FromClientToServerProperty (should be SetFromClient): {entity.FromClientToServerProperty}")
            Console.WriteLine($"Client side: entity.TwoWayProperty (should be SetBeforeException): {entity.TwoWayProperty}")

            Console.Write("Press any key to continue...")
            Console.ReadKey(True)
            Console.WriteLine()

            serverRemoteAgencyInstance.Dispose()
            clientRemoteAgencyInstance.Dispose()
        End Sub

        Private Shared Sub ServerRemoteAgencyInstance_ExceptionRedirected(sender As Object, e As ExceptionRedirectedEventArgs)
            Console.WriteLine(
                "Server side exception: " + vbCrLf +
                $"  Interface: {e.ServiceContractInterface.FullName}" + vbCrLf +
                $"  InstanceId: {e.InstanceId}" + vbCrLf +
                $"  AssetName: {e.AssetName}" + vbCrLf +
                $"  ExceptionType: {e.RedirectedException.GetType().FullName}" + vbCrLf +
                $"  ExceptionMessage: {e.RedirectedException.Message}")
        End Sub
    End Class

End Namespace
